### PR TITLE
tektoncd-cli: cleanup

### PR DESCRIPTION
**Pull request by 06kellyjac**
###### Motivation for this change

- use buildGoModule buildPhase
- let buildGoModule set platform
- add -s -w ldflags for smaller build (-5MB)
  - 43MB -> 38MB
- add checkPhase running unit tests
- install man pages in 1 line
- add changelog
- add @vdemeester as a maintainer

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by 06kellyjac** [comment]
Result of `nixpkgs-review pr 112755` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>tektoncd-cli</li>
  </ul>
</details>

File: pkgs/applications/networking/cluster/tektoncd-cli/default.nix [hunk 1]
```diff
@@ -13,35 +13,39 @@ buildGoModule rec {
...
+    "-w"
+    "-X github.com/tektoncd/cli/pkg/cmd/version.clientVersion=${version}"
+  ];
 
   nativeBuildInputs = [ installShellFiles ];
 
-  buildPhase = ''
-    make bin/tkn
+  checkPhase = ''
+    make test-unit
```
Comment (vdemeester) [inline pkgs/applications/networking/cluster/tektoncd-cli/default.nix:28]: 🎉

**Review by vdemeester — APPROVED** [review]
LGTM 🐯

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 112755` at 3f5d12f2 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li><a href="https://gist.github.com/849e6b241ca823c8be2b22114e31f715">tektoncd-cli</a></li></li>
  </ul>
</details>

File: pkgs/applications/networking/cluster/tektoncd-cli/default.nix (hunk 1 above)
Comment (vdemeester) [reply pkgs/applications/networking/cluster/tektoncd-cli/default.nix:28]: Although, I think relying on `go test` (that `buildGoModule` does automatically right ?) should work fine.

Comment (06kellyjac) [reply pkgs/applications/networking/cluster/tektoncd-cli/default.nix:28]: Leaving `checkPhase` to `buildGoModule` said no tests found as it only looks in `cmd/tkn`

```
running tests
?       github.com/tektoncd/cli/cmd/tkn [no test files]
```

I'll try some different `subPackages` options

Comment (06kellyjac) [reply pkgs/applications/networking/cluster/tektoncd-cli/default.nix:28]: If you don't set `subPackages`:

```
Building subPackage ./third_party/VENDOR-LICENSE/github.com/hashicorp/errwrap
main module (github.com/tektoncd/cli) does not contain package github.com/tektoncd/cli/third_party/VENDOR-LICENSE/github.com/hashicorp/errwrap
error: --- Error
```

Comment (06kellyjac) [reply pkgs/applications/networking/cluster/tektoncd-cli/default.nix:28]: `rm -rf third_party` lets it run pretty well. Hitting a couple failing tests with the version

Should I sed `Client version: dev` -> `Client version: 0.16.0` ?

Alternatively I could update the golden files then testing it matches the golden files, but that seems useless

```
=== RUN   TestGetVersions/empty_deployment_items
    version_test.go:231: assertion failed:
        --- expected
        +++ actual
        @@ -1,3 +1,3 @@
        -Client version: dev
        +Client version: 0.16.0
         Pipeline version: unknown, pipeline controller may be installed in another namespace please use tkn version -n {namespace}



        You can run 'go test . -test.update-golden' to automatically update testdata/TestGetVersions-empty_deployment_items.golden to the new expected value.'

=== RUN   TestGetVersions/deployment_only_with_pipeline_installed
    version_test.go:231: assertion failed:
        --- expected
        +++ actual
        @@ -1,3 +1,3 @@
        -Client version: dev
        +Client version: 0.16.0
         Pipeline version: v0.10.0



        You can run 'go test . -test.update-golden' to automatically update testdata/TestGetVersions-deployment_only_with_pipeline_installed.golden to the new expected value.'

=== RUN   TestGetVersions/deployment_with_pipeline_and_triggers_installed
    version_test.go:231: assertion failed:
        --- expected
        +++ actual
        @@ -1,3 +1,3 @@
        -Client version: dev
        +Client version: 0.16.0
         Pipeline version: v0.10.0
         Triggers version: v0.5.0


        You can run 'go test . -test.update-golden' to automatically update testdata/TestGetVersions-deployment_with_pipeline_and_triggers_installed.golden to the new expected value.'

=== RUN   TestGetVersions/deployment_with_pipeline_and_dashboard_installed
    version_test.go:231: assertion failed:
        --- expected
        +++ actual
        @@ -1,3 +1,3 @@
        -Client version: dev
        +Client version: 0.16.0
         Pipeline version: v0.10.0
         Dashboard version: v0.7.0


        You can run 'go test . -test.update-golden' to automatically update testdata/TestGetVersions-deployment_with_pipeline_and_dashboard_installed.golden to the new expected value.'

=== RUN   TestGetVersions/deployment_with_pipeline,_triggers_and_dashboard_installed
    version_test.go:231: assertion failed:
        --- expected
        +++ actual
        @@ -1,3 +1,3 @@
        -Client version: dev
        +Client version: 0.16.0
         Pipeline version: v0.10.0
         Triggers version: v0.5.0


        You can run 'go test . -test.update-golden' to automatically update testdata/TestGetVersions-deployment_with_pipeline,_triggers_and_dashboard_installed.golden to the new expected value.'

--- FAIL: TestGetVersions (0.00s)
    --- FAIL: TestGetVersions/empty_deployment_items (0.00s)
    --- FAIL: TestGetVersions/deployment_only_with_pipeline_installed (0.00s)
    --- FAIL: TestGetVersions/deployment_with_pipeline_and_triggers_installed (0.00s)
    --- FAIL: TestGetVersions/deployment_with_pipeline_and_dashboard_installed (0.00s)
    --- FAIL: TestGetVersions/deployment_with_pipeline,_triggers_and_dashboard_installed (0.00s)
```

Comment (06kellyjac) [reply pkgs/applications/networking/cluster/tektoncd-cli/default.nix:28]: That worked, all tests passed. I'll commit and push. LMK if there's a cleaner way to do this :)

Comment (06kellyjac) [reply pkgs/applications/networking/cluster/tektoncd-cli/default.nix:28]: Would it be worth `rm -rf` all the .go files in VENDOR-LICENSE in https://github.com/tektoncd/cli?
Should I post a bug report in https://github.com/google/go-license?
Or is including the .go files a requirement of the license? Looks like it's all just MPL-2.0 License though.

```
λ find ./third_party/VENDOR-LICENSE -type f -name "*.go"
./third_party/VENDOR-LICENSE/github.com/hashicorp/errwrap/errwrap.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/errwrap/errwrap_test.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/golang-lru/2q_test.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/golang-lru/2q.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/golang-lru/simplelru/lru.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/golang-lru/simplelru/lru_test.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/golang-lru/simplelru/lru_interface.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/golang-lru/lru.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/golang-lru/arc.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/golang-lru/doc.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/golang-lru/lru_test.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/golang-lru/arc_test.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/go-multierror/multierror_test.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/go-multierror/sort_test.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/go-multierror/prefix.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/go-multierror/flatten_test.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/go-multierror/append.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/go-multierror/group.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/go-multierror/multierror.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/go-multierror/format_test.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/go-multierror/sort.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/go-multierror/prefix_test.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/go-multierror/flatten.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/go-multierror/append_test.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/go-multierror/format.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/go-multierror/group_test.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/go-version/version_collection_test.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/go-version/constraint.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/go-version/version_collection.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/go-version/constraint_test.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/go-version/version.go
./third_party/VENDOR-LICENSE/github.com/hashicorp/go-version/version_test.go
```

Comment (vdemeester) [reply pkgs/applications/networking/cluster/tektoncd-cli/default.nix:28]: oh I didn't think of that 😅 (both for the subpackage and the tests that check version)

**Comment by vdemeester** [comment]
LGTM 🐯

File: pkgs/applications/networking/cluster/tektoncd-cli/default.nix (hunk 1 above)
Comment (vdemeester) [reply pkgs/applications/networking/cluster/tektoncd-cli/default.nix]: I think it looks good 😉

Comment (06kellyjac) [reply pkgs/applications/networking/cluster/tektoncd-cli/default.nix]: :+1: If these files are removed from https://github.com/tektoncd/cli at some point we can drop the line

**Comment by 06kellyjac** [comment]
Result of `nixpkgs-review pr 112755` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>tektoncd-cli</li>
  </ul>
</details>

File: pkgs/applications/networking/cluster/tektoncd-cli/default.nix [hunk 2]
```diff
@@ -13,35 +13,43 @@ buildGoModule rec {
...
+    "-X github.com/tektoncd/cli/pkg/cmd/version.clientVersion=${version}"
+  ];
 
   nativeBuildInputs = [ installShellFiles ];
 
-  buildPhase = ''
-    make bin/tkn
+  preBuild = ''
+    # Breaks build/check as go files are still included
+    rm -rf ./third_party/VENDOR-LICENSE
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/cluster/tektoncd-cli/default.nix:27]:
```suggestion
    rm -r ./third_party/VENDOR-LICENSE
```

File: pkgs/applications/networking/cluster/tektoncd-cli/default.nix [hunk 3]
```diff
@@ -13,35 +13,43 @@ buildGoModule rec {
...
-    install bin/tkn -Dt $out/bin
 
-    mkdir -p "$out/share/man/man1"
-    cp docs/man/man1/* "$out/share/man/man1"
+  preCheck = ''
+    sed -i "s/dev/${version}/" pkg/cmd/version/testdata/TestGetVersions-*.golden
+  '';
+
+  postInstall = ''
+    install -D -m 644 docs/man/man1/* -t $out/man/man1
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/cluster/tektoncd-cli/default.nix:36]: Can we use installManPage here?

Comment (06kellyjac) [reply pkgs/applications/networking/cluster/tektoncd-cli/default.nix:36]: I think that's worked
```
 result ⇒ /nix/store/sz1dx4w37g1f07prnvfmsf3d1lzxiqqk-tektoncd-cli-0.16.0
├──  bin
│  └──  tkn
└──  share
   ├──  bash-completion
   │  └──  completions
   │     └──  tkn.bash
   ├──  fish
   │  └──  vendor_completions.d
   │     └──  tkn.fish
   ├──  man
   │  └──  man1
   │     ├──  tkn-clustertask-delete.1.gz
   │     ├──  tkn-clustertask-describe.1.gz
```

Comment (06kellyjac) [reply pkgs/applications/networking/cluster/tektoncd-cli/default.nix:36]: Just tweaking before pushing

File: pkgs/applications/networking/cluster/tektoncd-cli/default.nix (hunk 1 above)
Comment (06kellyjac) [reply pkgs/applications/networking/cluster/tektoncd-cli/default.nix]: @vdemeester Not sure where else to ask but is it possible to change the release note of a merged PR before release?

File: pkgs/applications/networking/cluster/tektoncd-cli/default.nix [hunk 4]
```diff
@@ -13,35 +13,47 @@ buildGoModule rec {
...
+
+  preCheck = ''
+    # Change the golden files to match our desired version
+    sed -i "s/dev/${version}/" pkg/cmd/version/testdata/TestGetVersions-*.golden
   '';
 
   installPhase = ''
-    install bin/tkn -Dt $out/bin
+    # Only install tkn
+    install -D $GOPATH/bin/tkn $out/bin/tkn
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/cluster/tektoncd-cli/default.nix:38]: Not sure about the $GOPATH here. Why did you change that?

Comment (kalbasit) [reply pkgs/applications/networking/cluster/tektoncd-cli/default.nix:38]: I think we should use `subPackages` and use postInstall to install completions and man pages rather than overriding the installPhase IMO.

Comment (06kellyjac) [reply pkgs/applications/networking/cluster/tektoncd-cli/default.nix:38]: Well is there an easy way to filter out the relevant subPackages? cmd/tkn obviously but that has none of the tests

Comment (06kellyjac) [reply pkgs/applications/networking/cluster/tektoncd-cli/default.nix]: Ok I'm excluding `third_party` & `cmd/docs` so now I don't need to rm third_party or get rid of cmd/docs result

Comment (06kellyjac) [reply pkgs/applications/networking/cluster/tektoncd-cli/default.nix]: The interface for `excludePackages` is really uncomfortable. It'd be nice if it could take an array like `subPackages`

File: pkgs/applications/networking/cluster/tektoncd-cli/default.nix [hunk 5]
```diff
@@ -13,35 +13,42 @@ buildGoModule rec {
...
-    mkdir -p "$out/share/man/man1"
-    cp docs/man/man1/* "$out/share/man/man1"
+  postInstall = ''
+    installManPage docs/man/man1/*
 
     installShellCompletion --cmd tkn \
       --bash <($out/bin/tkn completion bash) \
       --fish <($out/bin/tkn completion fish) \
-      --zsh <($out/bin/tkn completion zsh)
+      --zsh  <($out/bin/tkn completion zsh)
```
Comment (kalbasit) [inline pkgs/applications/networking/cluster/tektoncd-cli/default.nix:40]: Please revert this change which adds an extra unnecessary space.

**Comment by Mic92** [comment]
Result of `nixpkgs-review pr 112755` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>tektoncd-cli</li>
  </ul>
</details>
